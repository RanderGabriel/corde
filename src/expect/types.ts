import {
  EmojiLike,
  MessageIdentifier,
  MessageEditedIdentifier,
  MessageEmbedLike,
  ChannelLocation,
  RoleIdentifier,
  TestFunctionType,
  CordeBotLike,
} from "../types";
import { ColorResolvable } from "discord.js";
import { Colors, RolePermission } from "../utils";

export interface ExpectTestBaseParams {
  cordeBot: CordeBotLike;
  command: string | number | boolean;
  isNot: boolean;
  timeout: number;
  isCascade: boolean;
  channelId?: string;
  guildId?: string;
}
export interface ExpectTestParams extends ExpectTestBaseParams {
  testName: string;
}

export type MayReturnMatch = Matches<any> | void;

/**
 * Defines all functions that can be used
 * to check a bot reaction of a command.
 *
 */
export interface MessageMatches<TReturn extends MayReturnMatch> {
  /**
   * Defines the message expected to be returned by a
   * command.
   *
   * @param expect A message returned by a bot after invoke a command
   * @since 1.0
   */
  toReturn(expect: bigint | boolean | number | string): TReturn;
  /**
   * Defines the message expected to be returned by a
   * command.
   *
   * @param expect A message returned by a bot after invoke a command
   * @since 1.0
   */
  toReturn(expect: MessageEmbedLike): TReturn;

  /**
   * Defines the message expected to be returned by a
   * command in a specific channel and guild(optional).
   *
   * @param expect A message returned by a bot after invoke a command
   * @param channelId Id of the channel where the message must be sent to.
   * @param guildId Id of the guild that contains the channel where the message must be sent to.
   * @since 3.0
   */
  toReturnInChannel(
    expect: bigint | boolean | number | string | MessageEmbedLike,
    channelId: string,
    guildId?: string,
  ): TReturn;

  /**
   * Defines the message expected to be returned by a
   * command in a specific channel and guild(optional).
   *
   * @param expect A message returned by a bot after invoke a command.
   * @param channelLocation Object with to holds the id of the channel and of the guild (optional).
   * @since 3.0
   */
  toReturnInChannel(
    expect: bigint | boolean | number | string | MessageEmbedLike,
    channelLocation: ChannelLocation,
  ): TReturn;

  /**
   * Defines [reactions](https://discordjs.guide/popular-topics/reactions.html#reacting-to-messages)
   * that must be added to the command message.
   *
   * @param emojis Single or list of reactions that must be added to a message.
   * It can be **emojis** or [custom emojis](https://support.discord.com/hc/en-us/articles/360036479811-Custom-Emojis).
   *
   * @example
   *
   * // Checks if the first message sent by the testing bot receives emojis in reactions.
   * expect("emoji").toAddReaction(['😄']);
   * expect("emoji").toAddReaction([{ name: '😄' }]);
   * expect("emoji").toAddReaction([{ name: '😄' }]);
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }]);
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }]);
   *
   * // Checks if a specific message receives emojis in reactions.
   * expect("emoji").toAddReaction(['😄'], { id: '96008815106887111' });
   * expect("emoji").toAddReaction([{ name: '😄' }], { id: '96008815106887111' });
   *
   * // This example will find for a message with content: 'message text'
   * expect("emoji").toAddReaction([{ name: '😄' }], { name: 'message text' });
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }], { id: 'message text' });
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }], { id: 'message text' });
   *
   * // Its also possible to search the message directly by it's id:
   * expect("emoji").toAddReaction(['😄'], '96008815106887111');
   *
   * @since 1.0
   */
  toAddReaction(emojis: string[]): TReturn;
  toAddReaction(emojis: EmojiLike[]): TReturn;
  toAddReaction(emojis: (string | EmojiLike)[]): TReturn;
  toAddReaction(emojis: EmojiLike[], messageIdentifier: string): TReturn;
  toAddReaction(emojis: EmojiLike[], messageIdentifier: MessageIdentifier): TReturn;
  toAddReaction(emojis: (string | EmojiLike)[], messageIdentifier: string): TReturn;
  toAddReaction(emojis: (string | EmojiLike)[], messageIdentifier: MessageIdentifier): TReturn;
  toAddReaction(emojis: string[], messageIdentifier: string): TReturn;
  toAddReaction(emojis: string[], messageIdentifier: MessageIdentifier): TReturn;

  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, "😆"]);
   * toRemoveReaction([{ id: "12321" }, "😆"], "31241212512");
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: EmojiLike[], messageId?: string): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, { name: "😆" }]);
   * toRemoveReaction([{ id: "12321" }, { name: "😆" }], { id: "1232142"});
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: EmojiLike[], messageIdentifier?: MessageIdentifier): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @param messageId Id of the message.
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, "😆"]);
   * toRemoveReaction([{ id: "12321" }, "😆"], "4121512121");
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: (string | EmojiLike)[], messageId?: string): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, "😆"]);
   * toRemoveReaction([{ id: "12321" }, "😆"], { id: "4121512121" });
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: (string | EmojiLike)[], messageIdentifier?: MessageIdentifier): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed
   * @param messageId Id of the message.
   * @example
   *
   * toRemoveReaction(["😆"]);
   * toRemoveReaction(["😆"], { id: "13124124" });
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: string[], messageId?: string): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed.
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @example
   *
   * toRemoveReaction(["😆", "🥱"]);
   * toRemoveReaction(["😆", "🥱"], { id: "1234871972" });
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: string[], messageIdentifier?: MessageIdentifier): TReturn;

  /**
   * Verify if a command pinned a message.
   *
   * @param messageId Id of the message
   * @since 2.0
   */
  toPin(messageId: string): TReturn;
  /**
   * Verify if a command pinned a message.
   *
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @since 2.0
   */
  toPin(messageIdentifier: MessageIdentifier): TReturn;

  /**
   * Verify if a command unpinned a message.
   *
   * @param messageId Id of the message
   * @since 2.0
   */
  toUnPin(messageId: string): TReturn;
  /**
   * Verify if a command unpinned a message.
   *
   * @param message Object with **id** or **content** of the message.
   * @since 2.0
   */
  toUnPin(messageIdentifier: MessageIdentifier): TReturn;

  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **string** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: string, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **string** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: string, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **number** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: number, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **number** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: number, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **bigint** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: bigint, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **bigint** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: bigint, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **boolean** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: boolean, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **boolean** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: boolean, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **MessageEmbedLike** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: MessageEmbedLike, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **MessageEmbedLike** value for the message.
   * @param messageId Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: MessageEmbedLike, messageId?: string): TReturn;
}

/**
 * Tests for a **Role** structure.
 */
export interface RoleMatches<TReturn extends MayReturnMatch> {
  /**
   * Check if a command changed a role color.
   *
   * @param color The new color for the role.
   * @param roleId Id of the role.
   * @since 2.0
   */
  toSetRoleColor(color: ColorResolvable, roleId: string): TReturn;
  /**
   * Check if a command changed a role color.
   *
   * @param color The new color for the role.
   * @param roleIdentifier Object with the **id** or the **name** of the role.
   * @since 2.0
   */
  toSetRoleColor(color: ColorResolvable, roleIdentifier: RoleIdentifier): TReturn;
  /**
   * Check if a command changed a role color.
   *
   * @param color Color enum in hexadecimal format.
   * @param roleId Object with the **id** or the **name** of the role.
   * @since 2.0
   */
  toSetRoleColor(color: Colors, roleId: string): TReturn;
  /**
   * Check if a command changed a role color.
   *
   * @param color Color enum in hexadecimal format.
   * @param roleId Object with the **id** or the **name** of the role.
   * @since 2.0
   */
  toSetRoleColor(color: Colors, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command deletes a role.
   *
   * @param roleId Id of the role.
   * @since 2.0
   */
  toDeleteRole(roleId: string): TReturn;

  /**
   * Check if a command deletes a role.
   *
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @since 2.0
   */
  toDeleteRole(roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command defines a role as mentionable.
   *
   * @param mentionable If the role can or can not be mentionable.
   * @param roleId Id of the role.
   * @since 2.0
   */
  toSetRoleMentionable(mentionable: boolean, roleId: string): TReturn;
  /**
   * Check if a command define a role as mentionable.
   *
   * @param mentionable If the role can or can not be mentionable.
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @since 2.0
   */
  toSetRoleMentionable(mentionable: boolean, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a commend defines a role as a hoist.
   *
   * @param hoist if this role is pinned in the user listing.
   * @param roleId Id of the role.
   *
   * @description Discord provides two methods of displaying roles; hoisted and standard.
   * The role hierarchy is visibly clear to server members; roles are sorted and displayed
   * based on which role is higher in the role management menu.
   *
   * However, in a standard configuration, users are sorted alphabetically, meaning someone
   * with the highest role will be sorted wherever their name exists in the alphabet.
   *
   * Source from [discord support](https://support.discord.com/hc/en-us/community/posts/360060076751-Un-hoisted-Role-Hierarchy).
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRoleHoist(hoist: boolean, roleId: string): TReturn;
  /**
   * Check if a commend define a role as hoist.
   *
   * @param hoist if this role is pinned in the user listing.
   * @param roleIdentifier Object with **id** or **name** of the role.
   *
   * @description Discord provides two methods of displaying roles; hoisted and standard.
   * The role hierarchy is visibly clear to server members; roles are sorted and displayed
   * based on which role is higher in the role management menu.
   *
   * However, in a standard configuration, users are sorted alphabetically, meaning someone
   * with the highest role will be sorted wherever their name exists in the alphabet.
   *
   * Source from [discord support](https://support.discord.com/hc/en-us/community/posts/360060076751-Un-hoisted-Role-Hierarchy).
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRoleHoist(hoist: boolean, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command renames a role.
   *
   * @param newName new name of the role
   * @param roleId Id of the role.
   * @since 2.0
   */
  toRenameRole(newName: string, roleId: string): TReturn;
  /**
   * Check if a command rename a role.
   *
   * @param newName new name of the role
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @since 2.0
   */
  toRenameRole(newName: string, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command changes a role's position.
   *
   * @param newPosition The new position of the role.
   * @param roleId Id of the role.
   *
   * @description Role's maximum value depends on the number of roles the guild Have.
   * So, if there are only 3 roles (including the default *everyone*), the maximum
   * position that a role can have is 2 (The count begins with 0, So: 0, 1, 2).
   *
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRolePosition(newPosition: number, roleId: string): TReturn;
  /**
   * Check if a command change a role's position.
   *
   * @param newPosition The new position of the role.
   * @param roleIdentifier Object with **id** or **name** of the role.
   *
   * @description Role's maximum value depends on the number of roles the guild has.
   * So, if there are only 3 roles (including the default *everyone*), the maximum
   * position that a role can have is 2 (The count begins with 0, So: 0, 1, 2).
   *
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRolePosition(newPosition: number, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command change the
   * [permissions](https://discord.com/developers/docs/topics/permissions#permissions-bitwise-permission-flags)
   * of a role.
   *
   * @param roleId Id of the role.
   * @param permissions List of permissions allowed by Discord.
   *
   * @see https://discord.com/developers/docs/topics/permissions#permissions
   * @since 2.0
   */
  toSetRolePermission(roleId: string, ...permissions: RolePermission[]): TReturn;
  /**
   * Check if a command change the
   * [permissions](https://discord.com/developers/docs/topics/permissions#permissions-bitwise-permission-flags)
   * of a role.
   *
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @param permissions List of permissions allowed by Discord.
   *
   * @see https://discord.com/developers/docs/topics/permissions#permissions
   * @since 2.0
   */
  toSetRolePermission(roleIdentifier: RoleIdentifier, ...permissions: RolePermission[]): TReturn;
}

export interface MacherContructorArgs {
  commandName: unknown;
  isNot?: boolean;
  channelId?: string;
  guildId?: string;
  isCascade?: boolean;
}

export interface CascadeMatch {
  todoInCascade(...tests: TestFunctionType[]): void;
}

/**
 * Defines all functions that can be used
 * to check a bot reaction of a command.
 */
export type Matches<TReturn extends MayReturnMatch> = MessageMatches<TReturn> &
  RoleMatches<TReturn>;

/**
 * Defines the initial value of expectations from
 * **command** function. It includes all matches and
 * the *not* statement. Witch will deny the executed match
 */
export interface IsNot<TMatchesResponse extends any> {
  /**
   * Defines that a command should **not** do something.
   * Use this if you can not precise what response a command will throw,
   * But know what it **can not** throw.
   */
  not: TMatchesResponse;
}

export interface SetChannelMatchers<TReturn extends MayReturnMatch> {
  /**
   * Specify a channel where tests will be **validated** on.
   *
   * @example
   *
   * // Suposing that the config is:
   * const cordeConfig = {
   *  channelId: "123",
   *  botPrefix: "!"
   * };
   *
   * expect("ping").inChannel("321").toReturn("pong");
   *
   * @description This will send the message "!ping" in the channel defined in configs("123"),
   * and check if a message with content "pong" will be sent to the text channel of if "321".
   *
   * @param id Id of the channel
   */
  inChannel(id: string): IsNot<MessageMatches<TReturn>> & MessageMatches<TReturn>;
}

export interface SetGuildMatchers<TReturn extends MayReturnMatch> {
  /**
   * Specify a guild where tests will be **validated** in.
   *
   * @example
   *
   * // Suposing that the config is:
   * const cordeConfig = {
   *  guildId: "123",
   *  botPrefix: "!"
   * };
   *
   * expect("ping").inGuild("321").toReturn("pong");
   *
   * @description This will send the message "!ping" in the channel defined in configs("123"),
   * and check if a message with content "pong" will be sent to the text channel of if "321".
   *
   * @param id Id of the channel
   */
  inGuild(id: string): IsNot<RoleMatches<TReturn>> & RoleMatches<TReturn>;
}

export type AllMatches<TReturn extends MayReturnMatch> = SetChannelMatchers<TReturn> &
  SetGuildMatchers<TReturn> &
  IsNot<Matches<TReturn>> &
  Matches<TReturn>;

export interface Expect extends AllMatches<any> {
  /**
   * Receives which command will be tested.
   *
   * Do not inform the command prefix if
   * it's already informed in **configs**
   *
   * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
   *
   * @returns An object with all possible tests to be done
   * in the bot.
   *
   * @since 1.0
   */
  <T extends (() => number | string) | number | string>(commandNameResolvable: T): AllMatches<void>;
}
