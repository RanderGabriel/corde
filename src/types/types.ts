import { ColorResolvable, Message, MessageEmbed } from "discord.js";
import { Colors, RolePermission } from "..";
import { CordeBot } from "../core";
import { EmbedFieldData, FileOptions, MessageAttachment } from "discord.js";
import { Stream } from "stream";

export type messageType = "text" | "embed";
export type messageOutputType = Message | MinifiedEmbedMessage;
export type messageExpectationType = string | MessageEmbed;
export type TestFunctionType = (cordeBot: CordeBot) => Promise<TestReport>;
export type VoidPromiseFunction = () => void | Promise<void>;
export type GenericFunction = (...args: any[]) => any;
export type Primitive = number | bigint | string | boolean;
export type ResolveFunction<TResult> = (value: TResult) => void;
export type RejectFunction = (reason?: any) => void;
export type EmojisType = string[] | EmojiLike[] | (string | EmojiLike)[];

/**
 * Get all function `T` parameters as they may be
 * optional.
 */
export type ParametersAsOptional<T extends GenericFunction> = Parameters<T> | void[];

/**
 * Available types of config files
 */
export type configFileType = "js" | "json" | "ts";

/**
 * Represents **command** structure
 */
export interface AssertionProps {
  commandName: string;
  expectation: messageExpectationType;
  usingTrueStatement: boolean;
  output?: messageOutputType;
  messageType: messageType;
}

/**
 * Contain all tests cases | groups of a test file.
 */
export interface TestFile {
  path: string;
  groups: Group[];
  isEmpty: boolean;
}

/**
 * Represents **test** structure
 */
export interface Test {
  name?: string | number | boolean;
  subTests?: Test[];
  testsFunctions: TestFunctionType[];
  testsReports?: TestReport[];
}

/**
 * Represents **group** structure
 */
export interface Group {
  name?: string | number | boolean;
  subGroups?: Group[];
  tests: Test[];
}

export interface Identifier {
  id?: string;
}

export interface RoleIdentifier extends Identifier {
  name?: string;
}

export interface BaseRole {
  name?: string;
  color?: ColorResolvable | Colors;
  isHoist?: boolean;
  position?: number;
  permissions?: RolePermission;
  isMentionable?: boolean;
}

/**
 * Contains a set of properties needed for execution of corde
 */
export interface ConfigOptions {
  /**
   * Fake bot used to test the really one
   */
  cordeTestToken?: string;
  /**
   * User's bot that will be tested
   */
  botTestId?: string;
  /**
   * User's bot token that will run.
   */
  botTestToken?: string;
  /**
   * Channel where tests will run
   */
  channelId?: string;
  /**
   * Guild where tests will run
   */
  guildId?: string;
  /**
   * Defines max amount of time that a command can run
   */
  timeOut?: number;
  /**
   * Defines how identify bot calls
   */
  botPrefix?: string;
  /**
   * Path for case tests. Use this from the base directory of the application
   */
  testFiles?: string[];
}

export interface Author {
  icon_url: string;
  name: string;
  url: string;
}

export interface Field {
  name: string;
  inline: boolean;
  value: string;
}

export interface Image {
  url: string;
}

export interface Thumbnail {
  url: string;
}

export interface MinifiedEmbedMessage {
  author: Author;
  color: number;
  description: string;
  fields: Field[];
  footer?: any;
  image: Image;
  thumbnail: Thumbnail;
  timestamp?: any;
  title: string;
  type: string;
  url: string;
}

export interface MessageIdentifier {
  /**
   * Text of a message, use it to find a message if you don't know
   * it's **id**.
   *
   * If there is more than one message with the same content,
   * Corde will handle the latest message sent.
   *
   * ps: To avoid possible inconsistences, recommend to use **id** for message search.
   */
  content?: string;
  /**
   * Identifier of the message
   */
  id?: string;
}

/**
 * Object contract used to identify messages in message
 * edition tests.
 */
export interface MessageEditedIdentifier {
  /**
   * Identifier of the message
   */
  id?: string;
  /**
   * Old content of the message to identify it.
   */
  oldContent?: string;
}

export interface TestReport {
  readonly pass: boolean;
  readonly message?: string;
  trace?: string;
}

export interface SemiRunnerReport {
  totalTests: number;
  totalEmptyTests: number;
  totalEmptyTestFiles: number;
  totalTestFiles: number;
  totalTestsPassed: number;
  totalTestsFailed: number;
  totalTestFilesPassed: number;
  totalTestFilesFailed: number;
}

export interface RunnerReport extends SemiRunnerReport {
  testTimer: string;
}

export interface MessageEmbedAuthor {
  name?: string;
  url?: string;
  iconURL?: string;
}

export interface MessageEmbedFooter {
  /**
   * footer text
   */
  text?: string;
  /**
   * url of footer icon (only supports http(s) and attachments)
   */
  iconURL?: string;
}

export interface MessageEmbedImage {
  /**
   * source url of image (only supports http(s) and attachments)
   */
  url: string;
  /**
   * height of image
   */
  height?: number;
  /**
   * width of image
   */
  width?: number;
}

export interface MessageEmbedThumbnail {
  /**
   * Url of the thumbnail
   */
  url: string;
  /**
   * Height of the thumbnail
   */
  height?: number;
  /**
   * width of the thumbnail
   */
  width?: number;
}

export interface FileLike {
  /**
   * Buffer, URL or stream of the file.
   *
   * @see https://nodejs.org/api/stream.html
   * @see https://nodejs.org/api/buffer.html
   */
  attachment: Buffer | string | Stream;
  /**
   * Name of the file
   */
  name: string;
}

/**
 * Main and optional informations about a embed message.
 */
export interface MessageEmbedLike {
  /**
   * author name **or** information
   */
  author?: MessageEmbedAuthor | string;
  /**
   * color code of the embed
   */
  color?: ColorResolvable;
  /**
   * description of embed
   */
  description?: string;
  /**
   * fields information. Array of embed field objects
   */
  fields?: EmbedFieldData[];
  /**
   * files urls **or** informations of the embed.
   */
  files?: (FileLike | string)[];
  /**
   * Footer url **or** information
   */
  footer?: MessageEmbedFooter | string;
  /**
   * Image URL **or** information
   */
  image?: MessageEmbedImage | string;
  /**
   * Source url of thumbnail (only supports http(s) and attachments)
   */
  thumbnailUrl?: string;
  /**
   * Timestamp of embed content **or** a Date object
   */
  timestamp?: number | Date;
  /**
   * Title of embed
   */
  title?: string;
  /**
   * Url of embed
   */
  url?: string;
}

export interface EmojiLike {
  id?: string;
  name?: string;
}
