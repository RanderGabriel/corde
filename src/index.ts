/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
