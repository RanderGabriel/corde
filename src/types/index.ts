export * from "./types";
export * from "./locale";
export * from "./discordTypes";
