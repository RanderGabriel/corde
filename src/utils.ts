import { ParameterNotFoundError } from './errors';

export function clone(object: object | object[]): object | object[] {
  if (!object) {
    throw new ParameterNotFoundError('Object can not be undefined');
  }

  if (object as object[]) {
    const copyArray = [];
    for (const i in object as []) {
      copyArray.push(copyObject((object as [])[i]));
    }
    return copyArray;
  } else {
    return copyObject(object);
  }
}

function copyObject(object: object) {
  return JSON.parse(JSON.stringify(object));
}
