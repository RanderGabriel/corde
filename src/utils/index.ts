export * from "./guid";
export * from "./colors";
export * from "./permission";
export * from "./executeWithTimeout";
export * from "./isFunction";
export * from "./bitfield";
export * from "./isPrimitiveValue";
export * from "./wait";
export * from "./pick";
export * from "./tryImport";
export * from "./formatObject";
export * from "./diff";
export * from "./isPromise";
export * from "./dateDiff";
export * from "./logUpdate";
export * from "./stringIsNullOrEmpty";
export * from "./timer";
