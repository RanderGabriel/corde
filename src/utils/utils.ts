import chalk from 'chalk';
import { ParameterNotFoundError } from '../errors';

/**
 * Clone a object using JSON.parse
 *
 * @param object Object to be cloned
 */
export function clone<T>(object: T | T[]): T | T[] {
  if (!object) {
    throw new ParameterNotFoundError('Object can not be undefined');
  }
  return JSON.parse(JSON.stringify(object));
}

/**
 * Pick some properties of a object
 *
 * @see https://www.typescriptlang.org/docs/handbook/utility-types.html#picktk
 * @param obj Object to get its properties
 * @param keys Properties that must be got
 */
export function pick<T, K extends keyof T>(obj: T, ...keys: K[]): Pick<T, K> {
  const copy = {} as Pick<T, K>;
  keys.forEach((key) => (copy[key] = obj[key]));
  return copy;
}

/**
 * Finish the process with code 1
 * displaying a default message in case of no data in parameter.
 *
 * @param msg Message to be displayed before exit the process
 */
export function exitProcessWithError(msg?: string) {
  if (msg) {
    console.log(`\n${chalk.red('error')} ${msg}`);
  } else {
    console.log(`\n${chalk.red('error')} Command failed with exit code 1`);
  }
  process.exit(1);
}
