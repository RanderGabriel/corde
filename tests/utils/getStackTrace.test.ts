import { TestReport } from "../../src/types";
import { getStackTrace } from "../../src/utils";

type TypeExecutor = () => Promise<TestReport>;
let executorList: TypeExecutor[] = [];

class ExpectExample {
  toReturn() {
    const trace = getStackTrace();
    executorList.push(() => {
      return Promise.resolve<TestReport>({
        pass: false,
        message: "error",
      }).then((r) => {
        r.trace = trace;
        return r;
      });
    });
  }
}

function test() {
  testExpect().toReturn();
}

export function testExpect() {
  return new ExpectExample();
}

describe("testing getStackTrace", () => {
  it("should get trace of an function", async () => {
    test();
    const report = await executorList[0]();
    expect(report).toMatchSnapshot();
  });
});
