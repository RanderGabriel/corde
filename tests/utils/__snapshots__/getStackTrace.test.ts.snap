// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing getStackTrace should get trace of an function 1`] = `
Object {
  "message": "error",
  "pass": false,
  "trace": "    test (tests\\\\utils\\\\getStackTrace.test.ts:29:16)
    Object.<anonymous> (tests\\\\utils\\\\getStackTrace.test.ts:39:5)",
}
`;
